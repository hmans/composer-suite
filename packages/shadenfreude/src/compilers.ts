import { RenderCallback } from "@react-three/fiber"
import { IUniform } from "three"
import { variablesToNodes } from "./factories"
import { formatValue } from "./formatters"
import { isVariable, ShaderNode, Value, Variable } from "./types"

type Program = "vertex" | "fragment"

export function compileVariableValue(value: Value): string {
  return isVariable(value) ? value.name : formatValue(value)
}

export function compileVariable(variable: Variable) {
  const valueString =
    variable.qualifier !== "uniform" && variable.value !== undefined
      ? ` = ${compileVariableValue(variable.value)}`
      : ""

  return `${variable.qualifier ?? ""} ${variable.type} ${
    variable.name
  }${valueString};`
}

function nodeTitle(node: ShaderNode) {
  return `/** Node: ${node.name} **/`
}

function dependencies(node: ShaderNode, deps = new Array<ShaderNode>()) {
  for (const variable of Object.values(node.inputs)) {
    if (isVariable(variable.value)) {
      /* get dependency */
      const dependency = variablesToNodes.get(variable.value)!
      if (!dependency) throw new Error("Dependency not found")

      /* If we haven't seen this dependency yet, invoke the callback */
      if (!deps.includes(dependency)) {
        dependencies(dependency, deps)
      }
    }
  }

  deps.push(node)
  return deps
}

function compileHeader(node: ShaderNode, program: Program) {
  const header = (n: ShaderNode) => `
    ${nodeTitle(n)}

    ${Object.values(n.uniforms)
      .map((variable) => compileVariable({ ...variable, qualifier: "uniform" }))
      .join("\n")}

    ${Object.values(n.varyings)
      .map((variable) => compileVariable({ ...variable, qualifier: "varying" }))
      .join("\n")}

    ${n[program].header ?? ""}
  `

  return dependencies(node)
    .map(header)
    .join("\n\n\n")
}

function compileBody(node: ShaderNode, program: Program) {
  return dependencies(node)
    .map(
      (node) => `
        ${nodeTitle(node)}

        ${Object.entries(node.outputs)
          .map(([_, variable]) =>
            compileVariable({ ...variable, value: undefined })
          )
          .join("")}

        {
          /* Varying References */
          ${Object.entries(node.varyings)
            .map(([name, variable]) =>
              compileVariable({ ...variable, name, value: variable.name })
            )
            .join("")}

          /* Inputs */
          ${Object.entries(node.inputs)
            .map(([name, variable]) => compileVariable({ ...variable, name }))
            .join("")}

          /* Outputs */
          ${Object.entries(node.outputs)
            .map(([name, variable]) => compileVariable({ ...variable, name }))
            .join("")}

          /* Code */
          ${node[program].body ?? ""}

          /* Update globals */
          ${Object.entries(node.outputs)
            .map(([name, variable]) => `${variable.name} = ${name};`)
            .join("\n")}
          ${
            program === "vertex"
              ? Object.entries(node.varyings)
                  .map(([name, variable]) => `${variable.name} = ${name};`)
                  .join("")
              : ""
          }
          }
      `
    )
    .join("\n\n\n")
}

export function compileShader(root: ShaderNode) {
  /* Build uniforms */
  const uniforms = {} as Record<string, IUniform>

  dependencies(root).forEach((node) => {
    Object.entries(node.uniforms).forEach(([name, variable]) => {
      uniforms[variable.name] = variable as IUniform
    })
  })

  /* Build callbacks */
  const callbacks = dependencies(root).map((node) => node.update)

  const update: RenderCallback = (...args) => {
    callbacks.forEach((callback) => callback?.(...args))
  }

  /* Build vertex shader program */
  const vertexShader = `
    /*** VERTEX SHADER ***/

    ${compileHeader(root, "vertex")}

    void main() {
      ${compileBody(root, "vertex")}
    }`

  /* Build fragment shader program */
  const fragmentShader = `
    /*** FRAGMENT SHADER ***/

    ${compileHeader(root, "fragment")}

    void main() {
      ${compileBody(root, "fragment")}
    }`

  return {
    vertexShader,
    fragmentShader,
    uniforms,
    update
  }
}
