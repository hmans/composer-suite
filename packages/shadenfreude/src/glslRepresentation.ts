import { Color, Vector2, Vector3, Vector4 } from "three"
import { isVariable, Value } from "./variables"

export const glslRepresentation = (value: Value): string => {
  if (isVariable(value)) return value.name

  if (typeof value === "string") return value

  if (typeof value === "boolean") return value ? "true" : "false"

  if (typeof value === "number") {
    const s = value.toString()
    return s.match(/[.e]/) ? s : `${s}.0`
  }

  if (value instanceof Color)
    return `
      vec3(
        ${glslRepresentation(value.r)},
        ${glslRepresentation(value.g)},
        ${glslRepresentation(value.b)}
      )`

  if (value instanceof Vector2)
    return `
      vec2(
        ${glslRepresentation(value.x)},
        ${glslRepresentation(value.y)}
      )`

  if (value instanceof Vector3)
    return `
      vec3(
        ${glslRepresentation(value.x)},
        ${glslRepresentation(value.y)},
        ${glslRepresentation(value.z)}
      )`

  if (value instanceof Vector4)
    return `
      vec4(
        ${glslRepresentation(value.x)},
        ${glslRepresentation(value.y)},
        ${glslRepresentation(value.z)}
        ${glslRepresentation(value.w)}
      )`

  /* TODO: Matrix3 */

  /* TODO: Matrix4 */

  /* Fail */
  throw new Error(`Could not render value to GLSL: ${value}`)
}
