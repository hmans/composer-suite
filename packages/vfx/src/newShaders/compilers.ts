import { formatValue } from "../shaders"
import { IVariable, Shader, VariableQualifier, Variables } from "./types"

function compileVariable(
  qualifier: VariableQualifier,
  name: string,
  { type, value }: IVariable
) {
  const valueString = value !== undefined ? ` = ${formatValue(value)}` : ""
  return `${qualifier} ${type} ${name}${valueString};`
}

function compileVariables(qualifier: VariableQualifier, variables: Variables) {
  return Object.entries(variables)
    .map(([name, variable]) => compileVariable(qualifier, name, variable))
    .join("\n")
}

function compileProgram(
  shader: Shader,
  program: "vertexShader" | "fragmentShader"
) {
  const {
    uniforms,
    varyings,
    [program]: { header, main }
  } = shader

  return `
    ${compileVariables("uniform", uniforms)}
    ${compileVariables("varying", varyings)}
    ${header}

    void main() {
      ${main}
    }`
}

export function compileMaterial(shader: Shader) {
  return {
    uniforms: shader.uniforms,
    vertexShader: compileProgram(shader, "vertexShader"),
    fragmentShader: compileProgram(shader, "fragmentShader")
  }
}
