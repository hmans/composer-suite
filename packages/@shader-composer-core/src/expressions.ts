import { glslRepresentation } from "./glslRepresentation"

const zip = (a: TemplateStringsArray, b: any[]) => a.map((k, i) => [k, b[i]])

export type Expression = {
  _: "Expression"
  values: any[]
  render: () => string
  toString: () => string
}

export const glsl = (strings: TemplateStringsArray, ...values: any[]): Expression => {
  const render = () =>
    zip(
      strings,
      values.map((v) => glslRepresentation(v))
    )
      .flat()
      .join("")

  return {
    _: "Expression",
    values,
    render,
    toString: render
  }
}

/** A shortcut for the `glsl` tagged template literal helper. */
export const $ = glsl

export function isExpression(v: any): v is Expression {
  return v && v._ === "Expression"
}
