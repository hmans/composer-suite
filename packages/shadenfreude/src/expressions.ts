import { glslRepresentation } from "./glslRepresentation"

export type Expression = {
  _: "Expression"
  values: any[]
  render: () => string
}

const zip = (a: TemplateStringsArray, b: any[]) => a.map((k, i) => [k, b[i]])

export const code = (
  strings: TemplateStringsArray,
  ...values: any[]
): Expression => ({
  _: "Expression",

  values,

  render: () =>
    zip(strings, values.map(glslRepresentation))
      .flat()
      .join("")
})

export function isExpression(v: any): v is Expression {
  return v && v._ === "Expression"
}
