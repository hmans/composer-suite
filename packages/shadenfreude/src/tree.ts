import { Color, Matrix3, Matrix4, Vector2, Vector3, Vector4 } from "three"
import { Expression } from "./expressions"
import { identifier, Part } from "./lib/concatenator3000"
import idGenerator from "./lib/idGenerator"

/**
 * The different GLSL types we're supporting in nodes.
 */
export type GLSLType =
  | "bool"
  | "float"
  | "vec2"
  | "vec3"
  | "vec4"
  | "mat3"
  | "mat4"

export type JSTypes = {
  bool: boolean
  float: number
  vec2: Vector2
  vec3: Vector3 | Color
  vec4: Vector4
  mat3: Matrix3
  mat4: Matrix4
}

export type Value<T extends GLSLType = any> = Expression | JSTypes[T] | Node<T>

export type Chunk = Part | Part[]

export type NodeConfig<T extends GLSLType = any> = {
  id: number
  title: string
  name: string
  only?: "vertex" | "fragment"
  varying?: boolean
  vertexHeader?: Chunk
  vertexBody?: Chunk
  fragmentHeader?: Chunk
  fragmentBody?: Chunk
}

/**
 * Any object that extends the INode type can be a node in the shader tree.
 * Node objects are free to expose any additional properties on top of this.
 */
export interface INode<T extends GLSLType = any> {
  _: "Node"
  _config: NodeConfig<T>
  type: T
  value: Value<T>
}

export type Node<
  T extends GLSLType = any,
  API extends Record<string, any> = {}
> = INode<T> & API

const nextAnonymousId = idGenerator()

/**
 * Create a node for the shader tree.
 *
 * @param type GLSL type of the node.
 * @param value Value of the node. Can be a JS value, a reference to another node, or a string expression.
 * @param configInput Optional configuration object.
 * @returns A freshly created node, just for you
 */
export const Node = <T extends GLSLType>(
  type: T,
  value: Value<T>,
  configInput: Partial<NodeConfig<T>> = {}
) => {
  const id = nextAnonymousId()

  const config: NodeConfig<T> = {
    /* Defaults */
    id,
    title: "anon",
    name: identifier("anonymous", id),

    /* User-provided configuration */
    ...configInput
  }

  const v: Node<T> = {
    _: "Node",
    _config: config,
    type,
    value
  }

  return v
}

export function isNode(v: any): v is Node {
  return v && v._ === "Node"
}

/* Helpers */

const makeNodeHelper = <T extends GLSLType>(type: T) => (
  v: Value<T>,
  extras?: Partial<NodeConfig<T>>
) => Node(type, v, extras)

export const Float = makeNodeHelper("float")
export const Bool = makeNodeHelper("bool")
export const Vec2 = makeNodeHelper("vec2")
export const Vec3 = makeNodeHelper("vec3")
export const Vec4 = makeNodeHelper("vec4")
export const Mat3 = makeNodeHelper("mat3")
export const Mat4 = makeNodeHelper("mat4")
