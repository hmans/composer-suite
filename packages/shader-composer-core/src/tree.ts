import { Expression, isExpression } from "./expressions"
import { isSnippet, Snippet } from "./snippets"
import { isUnit, Unit, Input, Program } from "./units"

export type Item = Input | Snippet | Expression

/**
 * Given a root unit, iterate over the tree and invoke the given callback for each
 * item encountered. Items include units, expressions, snippets, and any constant
 * values.
 *
 * @param item The root of the tree to traverse.
 * @param callback The callback to execute for each item.
 */
export const walkTree = (
  item: Item,
  program: Program | "any",
  callback: (item: Item) => void,
  seen = new Set<Item>()
) => {
  /* Visit each item only once */
  if (seen.has(item)) return
  seen.add(item)

  /* Dive into dependencies */
  for (const dependency of getDependencies(item, program)) {
    walkTree(dependency, program, callback, seen)
  }

  /* Invoke callback */
  callback(item)
}

/**
 * Walks the tree and returns all items found where the given callback function
 * returns true.
 */
export const collectFromTree = (
  root: Item,
  program: Program | "any",
  check?: (item: Item) => boolean
) => {
  const found = new Array<Item>()

  walkTree(root, program, (item) => {
    if (check !== undefined ? check(item) : true) {
      found.push(item)
    }
  })

  return found
}

/**
 * Given a unit, expression o snippet, returns that item's dependencies.
 *
 * @param item
 * @returns
 */
export const getDependencies = (
  item: Item,
  program: Program | "any"
): Item[] => {
  const dependencies = isUnit(item)
    ? getUnitDependencies(item, program)
    : isExpression(item)
    ? item.values
    : isSnippet(item)
    ? item.expression.values
    : []

  return dependencies.flat(Infinity).filter(dependencyFilter)
}

const getUnitDependencies = (
  { _unitConfig: config }: Unit,
  program: Program | "any"
) => {
  const dependencies = new Array<Item>()

  if (!config.varying || program === "any" || program === "vertex")
    dependencies.push(config.value)

  if (program === "any" || program === "vertex")
    dependencies.push(
      config.vertex?.header?.values,
      config.vertex?.body?.values
    )

  if (program === "any" || program === "fragment")
    dependencies.push(
      config.fragment?.header?.values,
      config.fragment?.body?.values
    )

  return dependencies.filter(dependencyFilter)
}

const dependencyFilter = (item: any) => !!item
