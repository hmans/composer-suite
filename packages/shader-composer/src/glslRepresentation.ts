import { isExpression } from "./expressions"
import { isArrayWithLength, isObjectWithKeys } from "./glslType"
import { isSnippet } from "./snippets"
import { GLSLType, Input, isUnit, uniformName } from "./units"

export const glslRepresentation = (
  value: Input | undefined,
  typeHint?: GLSLType
): string => {
  if (value === undefined) return ""

  if (isUnit(value))
    return value._unitConfig.uniform
      ? uniformName(value)
      : value._unitConfig.variableName

  if (isExpression(value)) return value.render()

  if (isSnippet(value)) return value.name

  if (typeof value === "string") return value

  if (typeof value === "boolean") return value ? "true" : "false"

  if (typeof value === "number") {
    const s = value.toString()
    return typeHint === "int" ? s : s.match(/[.e]/) ? s : `${s}.0`
  }

  /* Number arrays. If we don't clone them into extra arrays, this code
  will somehow get confused when they are TypedArrays. Have not been able
  to figure out why. */

  if (isArrayWithLength(value, 2))
    return `vec2(${[...value].map((n: Input) => g(n)).join(", ")})`

  if (isArrayWithLength(value, 3))
    return `vec3(${[...value].map((n: Input) => g(n)).join(", ")})`

  if (isArrayWithLength(value, 4))
    return `vec4(${[...value].map((n: Input) => g(n)).join(", ")})`

  if (isArrayWithLength(value, 9))
    return `mat3(${[...value].map((n: Input) => g(n)).join(", ")})`

  if (isArrayWithLength(value, 16))
    return `mat4(${[...value].map((n: Input) => g(n)).join(", ")})`

  if (isObjectWithKeys(value, "isMatrix3"))
    return `mat3(${value
      .toArray()
      .map((n: Input) => g(n))
      .join(", ")})`

  if (isObjectWithKeys(value, "isMatrix4"))
    return `mat4(${value
      .toArray()
      .map((n: Input) => g(n))
      .join(", ")})`

  if (isObjectWithKeys(value, "x", "y", "z", "w"))
    return `vec4(${g(value.x)}, ${g(value.y)}, ${g(value.z)}, ${g(value.w)})`

  if (isObjectWithKeys(value, "x", "y", "z"))
    return `vec3(${g(value.x)}, ${g(value.y)}, ${g(value.z)})`

  if (isObjectWithKeys(value, "r", "g", "b"))
    return `vec3(${g(value.r)}, ${g(value.g)}, ${g(value.b)})`

  if (isObjectWithKeys(value, "x", "y"))
    return `vec2(${g(value.x)}, ${g(value.y)})`

  /* Fail */
  throw new Error(`Could not render value to GLSL: ${JSON.stringify(value)}`)
}

const g = glslRepresentation
