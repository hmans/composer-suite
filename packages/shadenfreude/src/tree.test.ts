import { compileShader } from "./compilers"
import { code } from "./expressions"
import { glslRepresentation } from "./glslRepresentation"
import { Float, Value, Node } from "./tree"

const glsl = glslRepresentation

describe("Node", () => {
  it("creates a node of the specified type and value", () => {
    const v = Node("float", 1)
    expect(v.type).toBe("float")
    expect(v.value).toBe(1)
    expect(glsl(v.value)).toBe("1.0")
  })

  it("supports string values (which will be used as verbatim expressions)", () => {
    const v = Node("vec3", code`vec3(1.0, 1.0, 1.0)`)
    expect(glsl(v.value)).toBe("vec3(1.0, 1.0, 1.0)")
  })

  it("supports expression values", () => {
    const a = Float(1)
    const v = Node("vec3", code`vec3(${a}, 1.0, 1.0)`)
    expect(glsl(v.value)).toBe(`vec3(${a._config.name}, 1.0, 1.0)`)
  })

  it("allows nodes to directly reference other nodes", () => {
    const source = Float(1)
    const v = Float(source)
    expect(v.value).toBe(source)
    expect(glsl(v.value)).toBe(source._config.name)
  })

  it("supports a 'varying' flag that will automatically make it pass its data as a varying", () => {
    const v = Node(
      "float",
      code`1.0 + 2.0 + onlyAvailableInVertex.x`, // a value expression that can only work in a vertex shader
      {
        title: "A variable with a varying",
        varying: true,
        vertexBody: `value += 3.0;` // Modiyfing the value before it gets put into the varying
      }
    )

    const [c] = compileShader(v)

    expect(c.vertexShader).toMatchSnapshot()
    expect(c.fragmentShader).toMatchSnapshot()
  })

  it("supports constructing nodes through constructor functions", () => {
    const Double = (f: Value<"float">) => Float(code`(${f}) * 2.0`)
    const v = Double(1)
    expect(glsl(v.value)).toBe("(1.0) * 2.0")
  })

  it("constructor functions can pass string values as expressions", () => {
    const Double = (f: Value<"float">) => Float(code`(${f}) * 2.0`)
    const v = Double(code`5.0`)
    expect(glsl(v.value)).toBe(`(5.0) * 2.0`)
  })

  it("constructor functions can pass references to other nodes", () => {
    const Double = (f: Value<"float">) => Float(code`(${f}) * 2.0`)
    const a = Float(1)
    const v = Double(a)
    expect(glsl(v.value)).toBe(`(${a._config.name}) * 2.0`)
  })

  it("constructor functions can pass expression values to other nodes", () => {
    const Double = (f: Value<"float">) => Float(code`(${f}) * 2.0`)
    const a = Float(5)
    const v = Double(code`${a} + 5.0`)
    expect(glsl(v.value)).toBe(`(${a._config.name} + 5.0) * 2.0`)
  })
})
