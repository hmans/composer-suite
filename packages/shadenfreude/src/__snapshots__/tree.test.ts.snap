// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node supports a 'varying' flag that will automatically make it pass its data as a varying 1`] = `
"/*** BEGIN: A variable with a varying (1) ***/
varying float v_float_A_variable_with_a_varying_1;
/*** END: A variable with a varying (1) ***/

void main()
{
  /*** BEGIN: A variable with a varying (1) ***/
  float float_A_variable_with_a_varying_1;
  {
    float value = 1.0 + 2.0 + onlyAvailableInVertex.x;
    value += 3.0;
    float_A_variable_with_a_varying_1 = value;
    v_float_A_variable_with_a_varying_1 = value;
  }
  /*** END: A variable with a varying (1) ***/

}"
`;

exports[`Node supports a 'varying' flag that will automatically make it pass its data as a varying 2`] = `
"/*** BEGIN: A variable with a varying (1) ***/
varying float v_float_A_variable_with_a_varying_1;
/*** END: A variable with a varying (1) ***/

void main()
{
  /*** BEGIN: A variable with a varying (1) ***/
  float float_A_variable_with_a_varying_1;
  {
    float value = v_float_A_variable_with_a_varying_1;
    float_A_variable_with_a_varying_1 = value;
  }
  /*** END: A variable with a varying (1) ***/

}"
`;
